### PR TITLE
PAMv3 implementation

### PR DESCRIPTION
**Pull request by kleewho**
feat: Method grant_token has been added

grant_token allows generation of signed token with permissions for channels, channel groups and uuids

### CODE REVIEW

File: lib/pubnub/validators/grant_token.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      def validate_permissions!(arg, name)
+        return if arg.nil?
+
+        raise(
+          ArgumentError.new(object: self, message: ":#{name} has to be kind of Hash for grant token event."),
+          ":#{name} has to be kind of Hash for grant token event."
+        ) unless arg.is_a?(Hash)
+      end
+
+      def validate_all_permissions!
```
Comment (budgetpreneur) [inline lib/pubnub/validators/grant_token.rb:50]: I think SDK should not validate all permissions. Our acceptance tests has scenario to test the request to server and capture response for this case.

Comment (kleewho) [reply lib/pubnub/validators/grant_token.rb:50]: Yeah. You're probably right

Comment (kleewho) [reply lib/pubnub/validators/grant_token.rb:50]: Although all of acceptance scenarios are passing, so it's not creating any problems I think

Comment (budgetpreneur) [reply lib/pubnub/validators/grant_token.rb:50]: so validation not working and request pass-through to server?

Comment (kleewho) [reply lib/pubnub/validators/grant_token.rb:50]: I'll check it

Comment (kleewho) [reply lib/pubnub/validators/grant_token.rb:50]: Heh. That's funny. The test is working and the validation is working as well 😆 When we have this scenario that we're not granting anything we're actually sending these grants for one uuid: `Permissions: {:read => false, :write => false, :manage => false , :delete => false, :create => false, :get => false, :update => false, :join => false}`. So... the validation is ok, at least one of `uuids`, `channels` or `channel_groups` are not empty and on the other hand nothing has been granted and we can test the server response

Comment (kleewho) [reply lib/pubnub/validators/grant_token.rb:50]: I'll remove that validation

**Review by budgetpreneur — COMMENTED** [review]
LGTM. I will let @parfeon approve the PR, if his comments were addressed.

File: lib/pubnub/events/grant_token.rb [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        sum |= 32
+      end
+
+      if @update
+        sum |= 64
+      end
+
+      if @join
+        sum |= 128
+      end
```
Comment (parfeon) [inline lib/pubnub/events/grant_token.rb:191]: @kleewho all these can be much smaller:
```ruby
sum |= 1 if @read
sum |= 2 if @write
sum |= 4 if @manage
sum |= 8 if @delete
sum |= 16 if @create
sum |= 32 if @get
sum |= 64 if @update
sum |= 128 if @join
```

And pretty easy to read :)

**Review by parfeon — APPROVED** [review]
Mostly looks good.

Just two places maybe require some attention :)

File: lib/pubnub/events/grant_token.rb (hunk 2 above)
Comment (kleewho) [reply lib/pubnub/events/grant_token.rb:191]: definitely better

File: .github/workflows/run_acceptance_tests.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        if: always()
+        with:
+          name: main.xml
+          path: main.xml
+      - name: Expose beta report
+        uses: actions/upload-artifact@v2
+        if: always()
+        with:
+          name: beta.xml
+          path: beta.xml
```
[the remark this answers is not in the document]
Comment (kleewho) [reply .github/workflows/run_acceptance_tests.yml:49]: Sure. Not a problem for me. I wasn't sure what's the expected way to do it

**Review by parfeon — APPROVED** [review]
LGTM!

**Comment by kleewho** [comment]
@client-engineering-bot release

**Review by parfeon — APPROVED** [review]
LGTM!

**Review by bartk — APPROVED** [review]

**Pull request merged**